### ISSUE TITLE
Problem detecting changes on affected commands in gitlab ci 

### ISSUE THREAD

**Issue by Sylphide** [nrwl/nx#8008]
Hello everyone,

I am not entirely sure if I did something wrong or if it is a real problem but here it is:

When running affected commands on gitlab CI it doesn't seem to do anything.

For example: I am currently on a feature branch where I commited some changes on an app in my monorepo
Locally I am able to use the command `yarn affected:test` and it works as intended running test on the changed app only.

However, on my gitlab CI, every affected commands just finishes in less than a second and no logs appears.

Here is my gitlab job script:

```
default:
  before_script:
    - apk add --no-cache git
    - git fetch origin master:master
    - git diff origin/master HEAD

build:
  stage: build
  script:
    - yarn --frozen-lockfile
    - yarn nx print-affected --base=origin/master --head=HEAD --verbose
    - yarn affected:build --prod --with-deps --base=origin/master --head=HEAD --parallel --skip-nx-cache --verbose
    - yarn affected:build --prod --with-deps --base=HEAD~1 --head=HEAD --parallel --skip-nx-cache --verbose
    - yarn nx run api:build:production
  artifacts:
    paths:
      - node_modules/
```

Here is the runner logs

```
$ yarn nx print-affected --base=origin/master --head=HEAD
yarn run v1.22.15
$ nx print-affected --base=origin/master --head=HEAD
Done in 4.17s.
$ yarn affected:build --prod --base=origin/master --head=HEAD --parallel --skip-nx-cache --verbose
yarn run v1.22.15
$ nx affected:build --prod --base=origin/master --head=HEAD --parallel --skip-nx-cache --verbose
Done in 0.77s.
$ yarn affected:build --prod --base=HEAD~1 --head=HEAD --parallel --skip-nx-cache --verbose
yarn run v1.22.15
$ nx affected:build --prod '--base=HEAD~1' --head=HEAD --parallel --skip-nx-cache --verbose
Done in 0.78s.
Uploading artifacts for successful job
```

Note that I tried various way of running my build command on affected projects. I tried to disable the default fixed depth of gitlab shallow cloning but it didn't do the job unfortunately. I also needed to fetch master through `git fetch origin master:master` otherwise master would not be recognized. I also printed a git diff in the runner and it shows correct changes.

I am relatively new to the nx ecosystem so it totally can be a misunderstanding on my part though.

Thanks for your help.

Edit:

In fact it's worse than that. I might have done something that made nx commands not work anymore in the CI.

Even a simple `npx nx run api:test` doesn't do anything anymore. If I want to run my tests on the CI I have to do `npx jest apps/api -c apps/api/jest.config.js`

**Comment by meeroslav** [comment]
Your `base` param is most likely not correct. Base refers to branch name or commit against which you are comparing to.
```bash
yarn affected:build --prod --base=master --head=HEAD
yarn affected:build --prod --base=4d2d3c5f693dced027e6f80931850fe096b768c0 --head=HEAD
```

For `PR` this should be the result of `git merge-base origin/master HEAD` - the moment in `master` history from which you branched out. If you set `base` to be just `master` you might end up with some odd affected results.

For `commit to main branch` base is usually `HEAD~1` - the previous commit, otherwise your base and your head will point to the same commit. This might explain what is happening to you.

The reason why it works on your local machine, is that affected is checking changed, uncommited and untracked files first, so most likely your `affected` was locally never reaching the point where `base` and `head` were important as seen in https://github.com/nrwl/nx/blob/master/packages/workspace/src/command-line/shared.ts#L8-L34

**Bonus**:
Sometimes, builds can fail on CI for various reasons. To avoid skipping silently carrying breaking changes further, you might want to compare top commit against last successful CI build (as opposed to just last one). This will allow you to always include broken commits when running build.

Unfortunately, we don't have automated script for GitLab that does this for you, but you can check the code from the GitHub action that does this:
https://github.com/nrwl/nx-set-shas/blob/main/find-successful-workflow.js

**Comment by Sylphide** [comment]
Thank you for your response, I just resolved my problem by cleaning up my dependencies. I think there was a version mismatch between my different @nrwl plugins making every nx commands fail silently. It was happening even in my local environment after reinstalling deps. I ended up removing my lockfile and my node_modules and after reinstalling deps I did an `nx migrate latest`. After that everything worked fine even on the CI. Weird though ¯\_(ツ)_/¯.
